src/pepr/operator/controllers/keycloak/authservice/config.ts

```diff
@@ -94,47 +120,122 @@ function buildInitialSecret(): AuthserviceConfig {
+ */
 export async function updateAuthServiceSecret(
   authserviceConfig: AuthserviceConfig,
   checksum = true,
-) {
-  const config = btoa(JSON.stringify(authserviceConfig));
-  const configHash = createHash("sha256").update(config).digest("hex");
+): Promise<void> {
+  return new Promise((resolve, reject) => {
+    // Update the in-memory secret immediately
+    inMemorySecret = authserviceConfig;
 
-  try {
-    // write the authservice config to the secret
-    await K8s(kind.Secret).Apply(
-      {
-        metadata: {
-          namespace: operatorConfig.namespace,
-          name: operatorConfig.secretName,
-        },
-        data: {
-          "config.json": config,
-        },
-      },
-      { force: true },
-    );
-  } catch (e) {
-    log.error(e, `Failed to write authservice secret`);
-    throw new Error("Failed to write authservice secret", { cause: e });
-  }
+    // Add the package config and its resolve function to the pending packages map
+    pendingPackages.set(authserviceConfig, resolve);
 
-  log.info("Updated authservice secret successfully");
+    // Clear the previous debounce timer, if it exists
+    if (debounceTimer) {
+      clearTimeout(debounceTimer);
+    }
 
-  if (checksum) {
-    log.info("Adding checksum to deployment authservice secret successfully");
-    await checksumDeployment(configHash);
-  }
+    // Set a new debounce timer to apply the update after the delay
+    debounceTimer = setTimeout(async () => {
+      try {
+        log.info(
+          `Applying debounced secret update for packages: ${Array.from(pendingPackages.keys()).length} pending packages`,
+        );
+
+        // Prepare the config to be written (assumes that all packages share the same secret)
+        const config = btoa(JSON.stringify(inMemorySecret));
+        const configHash = createHash("sha256").update(config).digest("hex");
+
+        // Write the in-memory authservice config to the Kubernetes secret
+        await K8s(kind.Secret).Apply(
+          {
+            metadata: {
+              namespace: operatorConfig.namespace,
+              name: operatorConfig.secretName,
+            },
+            data: {
+              "config.json": config,
+            },
+          },
+          { force: true },
+        );
+
+        log.info(`Updated authservice secret successfully for all pending packages.`);
+
+        // Apply the checksum if required
+        if (checksum) {
+          log.info(`Adding checksum to deployment for authservice secret`);
+          await checksumDeployment(configHash);
+        }
+
+        // Resolve the promises for all pending packages after the secret update
+        pendingPackages.forEach(resolveFunc => {
+          resolveFunc();
+        });
+      } catch (e) {
+        log.error(e, `Failed to write authservice secret`);
+
+        // Rollback to the last known successful state if the update fails
+        inMemorySecret = lastSuccessfulSecret;
```

Comment:
It looks like this is reverting back to `lastSuccessfulSecret` but that is only ever set/updated on initial secret fetch?

Reply:
@rjferguson21 iirc my though process here was that the lastSuccessfulSecret is global in memory state meaning it would be the original secret until a successful secret update at which time it would update, regardless if it was current thread or another thread, that way if one of the threads did fail it would still have the most current successfulSecret ( in the event there was or wasnt updates to the secret in the time it took to fail ).